daemon/k8s_watcher.go

```diff
@@ -1632,3 +1675,120 @@ func (d *Daemon) deleteK8sPodV1(pod *v1.Pod) {
 	}
 
 }
+
+func (d *Daemon) updateK8sNodeTunneling(k8sNodeOld, k8sNodeNew *v1.Node) error {
+	logger := log.WithFields(logrus.Fields{
+		"K8sNodeName": k8sNodeNew.ObjectMeta.Name,
+	})
+
+	nodeNew := k8s.ParseNode(k8sNodeNew)
+	// Ignore own node
+	if nodeNew.Name == node.GetName() {
+		return nil
+	}
+
+	getIDs := func(node *node.Node) (*identity.IPIdentityPair, string, error) {
+		if node == nil {
+			return nil, "", nil
+		}
+		hostIP := node.GetNodeIP(false)
+		if ip4 := hostIP.To4(); ip4 == nil {
+			logger.Debugf("Skipping ipcache update, HostIP is not an IPv4 address %s", hostIP)
+			return nil, "", nil
+		}
+
+		ciliumIPStr := k8sNodeNew.GetAnnotations()[annotation.CiliumHostIP]
+		ciliumIP := net.ParseIP(ciliumIPStr)
+		if ciliumIP == nil {
+			return nil, "", fmt.Errorf("no/invalid Cilium-Host IP: %s", ciliumIPStr)
+		}
+
+		id := &identity.IPIdentityPair{
+			IP:     ciliumIP,
+			HostIP: hostIP,
+			ID:     identity.ReservedIdentityCluster,
+		}
+		return id, ciliumIPStr, nil
+	}
+
+	identityPairNew, ciliumIPStrNew, err := getIDs(nodeNew)
+	if err != nil {
+		return err
+	}
+
+	var identityPairOld *identity.IPIdentityPair
+	if k8sNodeOld != nil {
+		nodeOld := k8s.ParseNode(k8sNodeOld)
+		var (
+			ciliumIPStrOld string
+			err            error
+		)
+		identityPairOld, ciliumIPStrOld, err = getIDs(nodeOld)
+		if err != nil {
+			return err
+		}
+
+		if identityPairOld != nil {
+			if identityPairOld.IP.Equal(identityPairNew.IP) &&
+				identityPairOld.HostIP.Equal(identityPairNew.HostIP) {
+				// If both identities are the same, there's no point on
+				// continuing doing any operation.
+				return nil
+			}
+
+			// Delete previous cilium IP for the given node.
+			ipcache.IPIdentityCache.Delete(ciliumIPStrOld)
+		}
+	}
+
+	updated := ipcache.IPIdentityCache.Upsert(ciliumIPStrNew, ipcache.Identity{
+		ID:     identity.ReservedIdentityCluster,
```

Comment:
Oh, interesting. Up until now, cluster has meant exactly one thing: On egress, the destination lies within the "cluster prefix" (typically 10.0.0.0/8) and is not otherwise managed by Cilium.

This now gives new meaning to cluster, to say it can also mean other nodes where Cilium is operating. Previous proposals [#3536, #4779] have proposed a new identity for this. I think we would benefit from actually having a new, different identity for this rather than making the meaning of "cluster" more complicated (unless you have a strong argument for what "cluster" should mean in future releases).

Note that this also introduces a potential policy incompatibility: IPs with the same role in Cilium 1.1 will now be classified as a different identity in Cilium 1.2. Therefore we should consider a `host-allows-world` style option for backwards compatibility. [host-allows-world](https://cilium.link/host-vs-world)

Reply:
For reference, currently, `reserved:cluster` matches the agent CLI `--ipv4-node` IP using the agent CLI `--ipv4-cluster-cidr-mask-size` CIDR range, ie by default it means `10.0.0.0/8` traffic and nothing else.

Reply:
We can map it to ReservedHost for now as it is via the kvstore and then discuss introducing a new identity in a separate PR.